### PR TITLE
Implement diff command

### LINKED ISSUE #58: Feature request: ability to diff cruft
Hi,

First of, thanks for making this fantastic project. It's a very nice improvement over cookiecutter and I'm relieved somebody tackled this problem.

The only thing I'm missing would be a way to check whether there's a diff between the remote template and the local one. If I'm not mistaken, the `check` command only checks for new commits in the remote which is not exactly what I want. My main use case is to ensure any improvement to the boilerplate is actually committed to the template and not to the local project, so that all other projects can benefit from the improvement.

Ideally (that may be another feature), if some projects using a cruft needs to apply some very specific changes to the template, they can either skip the whole files (which is already possible) or specify a set of patches to apply to the generated template. The patches could be specified in the `cruft.json` like so: `{"patches": [".cruftpatches/remove-useless-feature.patch]}`.

WDYT? I may be able to provide a PR with both features if required. Thanks.

### PR DESCRIPTION
**Pull request by gilbsgilbs**
This PR adds a new "diff" command that allows to compare the local
project against the remote template. Should the user make any change to
the remote template locally, it will appear when running "cruft diff".
This allows the user to quickly check how much the local boilerplate has
diverged from the template. Note that skipped and deleted files are
always ignored from the diff.

The diff command optionally takes an "--exit-code" flag that will
make sure cruft exits with status 1 should a non-empty diff exists.
This opens the door to stricter linting rules in CI. On top of that,
because this flag will likely be used in headless scripts, setting this
flag will disable the use of git's pager and coloration and will just
output the diff to standard output. This is very similar to "git diff"
behavior.

fixes #58

### CODE REVIEW

**Comment by gilbsgilbs** [comment]
FWIW, the coverage decreased because the test templates at https://github.com/samj1912/cookiecutter-test/ are too simple. I'd need a nested structure with multiple file to cover those two lines.

**Comment by samj1912** [comment]
Thanks for the PR! I will move the test cookiecutter to this organization so that contributors can submit changes to it to reflect test cases in cruft. I will have time to do it this weekend

File: tests/test_api.py [hunk 1]
```diff
@@ -120,3 +122,92 @@ def test_directory_and_checkout(tmpdir):
...
+        # without reprocessing by cruft. This delegates diff coloring and paging to git which
+        # improves user experience. As far as I know, there is no way to ask git diff to not
+        # display this path.
+        assert set(match.groups()) == {""}
+
+
+@pytest.mark.parametrize("exit_code", [(False,), (True,)])
+def test_diff_no_diff(exit_code, capfd, mocker, tmpdir):
+    project_dir = cruft.create(
+        # FIXME change this to samj1912 repo when it is merged.
```
Comment (gilbsgilbs) [inline tests/test_api.py:199]: reminder

File: tests/test_api.py [hunk 2]
```diff
@@ -120,3 +122,92 @@ def test_directory_and_checkout(tmpdir):
...
+    ],
+)
+def test_diff_has_diff(
+    exit_code, isatty, expect_reproducible_diff, expected_return_value, capfd, mocker, tmpdir
+):
+    mocker.patch.object(sys.stdout, "isatty", return_value=isatty)
+
+    project_dir = cruft.create(
+        # FIXME change this to samj1912 repo when it is merged.
+        "https://github.com/gilbsgilbs/cookiecutter-test",
```
Comment (gilbsgilbs) [inline tests/test_api.py:143]: reminder

**Comment by gilbsgilbs** [comment]
I've tested the final edge-cases and merged latest changes on master. Currently, the template is in my own fork, but I'll update the location in this PR as soon as the branch is created your own repo (see https://github.com/samj1912/cookiecutter-test/pull/1).

Cheers.

File: cruft/_cli.py [hunk 3]
```diff
@@ -231,3 +231,19 @@ def update(
...
     ):
         raise typer.Exit(1)
+
+
+@app.command(
+    short_help="Show the diff between the project and the current cruft template.",
+    help=_get_help_string(_commands.diff),
+)
+def diff(
+    project_dir: Path = typer.Option(
```
Comment (samj1912) [inline cruft/_cli.py:241]: Can we replicate all of the flags from cruft check here?

File: cruft/_commands/_generate.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import json
```
Comment (samj1912) [inline cruft/_commands/_generate.py:1]: Can you please create a separate commit with the common parts that have been moved? It will make it easier to review. So the way I'd do it is this. Retractor out the common parts from update in a separate commit. Then once that is done, implement the diff command in a separate commit.

File: cruft/_commands/_generate.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+CruftState = Dict[str, Any]
+
+
+def cookiecutter_template(
+    output_dir: Path,
+    repo: Repo,
+    project_dir: Path = Path("."),
+    cookiecutter_input: bool = False,
+    checkout: Optional[str] = None,
+) -> CCContext:
```
Comment (samj1912) [inline cruft/_commands/_generate.py:27]: Can we instead rename this to CookiecutterContext here?

File: cruft/_commands/diff.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                destination.mkdir(parents=True, exist_ok=True)
+                destination.chmod(local_path.stat().st_mode)
+
+        # Finally we can compute and print the diff.
+        diff = _diff.get_diff(remote_template_dir, local_template_dir)
+
+        if diff.strip():
+            has_diff = True
+
+            if exit_code or not sys.stdout.isatty():
```
Comment (samj1912) [inline cruft/_commands/diff.py:57]: git diff should already take care of this by itself? Do we still need to check for isatty?

File: cruft/_commands/update.py [hunk 7]
```diff
@@ -1,29 +1,14 @@
...
-    get_cruft_file,
-    is_project_updated,
-    json_dumps,
-)
-
-try:
-    import toml  # type: ignore
-except ImportError:  # pragma: no cover
-    toml = None  # type: ignore
+from . import _diff, _generate
```
Comment (samj1912) [inline cruft/_commands/update.py:10]: Can we move this under different sub packages in utils?

File: tests/test_cli.py [hunk 8]
```diff
@@ -19,20 +19,18 @@ def cruft_runner():
 
 @pytest.fixture
 def cookiecutter_dir(tmpdir):
```
Comment (samj1912) [inline tests/test_cli.py:21]: Why were these changed?

File: tests/test_cli.py [hunk 9]
```diff
@@ -148,8 +146,7 @@ def test_update(cruft_runner, cookiecutter_dir):
 
 
 def test_update_with_conflicts(cruft_runner, cookiecutter_dir):
-    with (cookiecutter_dir / "README.md").open("w") as f:
-        f.write("conflicts")
+    (cookiecutter_dir / "README.md").write_text("conflicts")
```
Comment (samj1912) [inline tests/test_cli.py:149]: Same here?

Let's try and limit the changes to the actual implementation in this PR and do the minor cosmetic updates in a separate one. This also applies to the other such changes that have been made.

Makes it easier to review and keeps the PR focused on 1 feature.

**Review by samj1912 — CHANGES_REQUESTED** [review]
Few minor comments, but it looks good overall! Thanks for the PR! Can you also update all the relevant docs in Readme and other places?

File: cruft/_commands/diff.py (hunk 6 above)
Comment (gilbsgilbs) [reply cruft/_commands/diff.py:57]: > Do we still need to check for isatty?

Yes, we need it because running `git diff` directly prefixes paths with `/tmp/xyz`. If the terminal isn't a TTY, we don't want this. (e.g. doing `cruft diff > cruft.patch` to apply the patch later on or make changes to it)

Comment (gilbsgilbs) [reply cruft/_commands/diff.py:57]: I'll add a comment to make it clear.

File: cruft/_cli.py (hunk 3 above)
Comment (gilbsgilbs) [reply cruft/_cli.py:241]: I'm not sure if this was a question or a request. `cruft check` command has 3 options:
- `--project-dir` that is already replicated in `cruft diff`
- `--checkout` that allows to specify against which branch to update
- `--strict` that I don't understand what it does but seems tightly related to the update process

So I'm not sure what I else could replicate here. In the meantime, I kind of like that `cruft diff` just diffs the local template against the remote one, following what's in the current `.cruft.json` and regardless of any update process. It just makes it more predictable and easy to reason about IMHO.

Tell me if you disagree and what you think I should replicate.

File: cruft/_commands/_generate.py (hunk 4 above)
Comment (gilbsgilbs) [reply cruft/_commands/_generate.py]: Done.

File: cruft/_commands/_generate.py (hunk 5 above)
Comment (gilbsgilbs) [reply cruft/_commands/_generate.py]: Done.

File: cruft/_commands/diff.py (hunk 6 above)
Comment (gilbsgilbs) [reply cruft/_commands/diff.py:56]: Done.

File: cruft/_commands/update.py (hunk 7 above)
Comment (gilbsgilbs) [reply cruft/_commands/update.py]: Done.

File: tests/test_cli.py (hunk 8 above)
Comment (gilbsgilbs) [reply tests/test_cli.py]: I removed them.

File: tests/test_cli.py (hunk 9 above)
Comment (gilbsgilbs) [reply tests/test_cli.py]: I removed them.

File: cruft/_cli.py (hunk 3 above)
Comment (samj1912) [reply cruft/_cli.py:241]: I think checkout would definitely be useful here.

Comment (gilbsgilbs) [reply cruft/_cli.py:241]: I see. Should be quite easy to implement.

File: cruft/_commands/utils/generate.py [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+    with TemporaryDirectory() as tmpdir_:
+        tmpdir = Path(tmpdir_)
+
+        # Kindly ask cookiecutter to generate the template
+        template_dir = generate_files(
+            repo_dir=inner_dir, context=new_context, overwrite_if_exists=True, output_dir=tmpdir
+        )
+        template_dir = Path(template_dir)
+
+        # Move the template content to the output directory
```
Comment (samj1912) [inline cruft/_commands/utils/generate.py:83]: We should not do this. This breaks the expected behavior for the cruft create command. Again similar to my other requests, let's please avoid any changes to other parts of the code expect the refactoring that's needed for diff to work.

**Review by samj1912 — CHANGES_REQUESTED** [review]

File: cruft/_cli.py [hunk 11]
```diff
@@ -231,3 +231,19 @@ def update(
...
+        Path("."), "--project-dir", "-p", help="Path to the project directory.", show_default=False
+    ),
+    exit_code: bool = typer.Option(
```
Comment (samj1912) [inline cruft/_cli.py:244]: Let's call this the check flag. it's more in line with other liners

File: cruft/_commands/utils/generate.py (hunk 10 above)
Comment (gilbsgilbs) [reply cruft/_commands/utils/generate.py:83]: That's actually useful for contributors like me to be able to reason on the code. I'm ok with making self-contained PRs and everything, but I prefer refactoring an implementation when the existing one would make me write convoluted code. And the current implementation of this on master would have made me write convoluted code, that's why I changed it.

Rather than shaming and looking down on contributors (friendly reminder that I also do this in my spare time, a sunday night, because I like your project), you could try to 1) understand why I did it 2) be precise on what exactly it does break and 3) focus on how we could not-break it while keeping it understandable to human beings.

File: cruft/_cli.py (hunk 11 above)
Comment (gilbsgilbs) [reply cruft/_cli.py:244]: More in line with other linters, less in line with git. But who uses git anyways…

Comment (gilbsgilbs) [reply cruft/_cli.py:244]: One argument to keep it as `--exit-code`: `--check` natural shortcode is `-c`, but that would conflict with `--checkout`. If you want me to change it, give me the shortcode you want for `--check`.

File: cruft/_commands/utils/generate.py (hunk 10 above)
Comment (samj1912) [reply cruft/_commands/utils/generate.py:83]: I'm sorry if you got that impression in anyway from my comments. (I even thanked you for the pull request :)) I'm not trying to look down or shame anyone here. I'm not sure what made think that way. I have tried to be cordial and I am extremely sorry if you got this impression.

As for my review - 

This currently breaks the cruft create command as the output directory is no longer determined by the cookiecutter templated directory name. Although this is a minor refactoring for diff and update, this is a major change for users using cruft to create templates. 

This is the same reason why it wasn't just implemented in the upstream pull request directly and instead implemented behind a strip flag.

Lastly, I think your reminder also applies to me, this is not my full time job and I'm doing my best to review this as quickly as possible in my spare time on a Sunday :) 

I also don't have any expectations from you to get back to me with the changes from my review instantly. If it's late, you can get back to this pull request later when you get a chance.

Again I completely appreciate your time and contributions. Please let me know if you need any clarity on any of the other comments I have made.

Comment (samj1912) [reply cruft/_commands/utils/generate.py:83]: Edit - I saw the code again, sorry my mistake. Looks like it's different function calls. Please ignore this review comment. It makes sense :)

File: tests/test_api.py (hunk 1 above)
Comment (samj1912) [reply tests/test_api.py:204]: The upstream branch has been created at https://github.com/cruft/cookiecutter-test/tree/diff

File: cruft/_cli.py (hunk 11 above)
Comment (samj1912) [reply cruft/_cli.py:244]: Thanks for the explanation. That makes sense. 

PS - I slightly edited your previous comment about git to be more in line with our [CoC](https://cruft.github.io/cruft/docs/contributing/3.-code-of-conduct/). I would again like to apologize if my review comments  made you feel like I was being inconsiderate, that was not my intention, but let's please keep the discussion polite. We are all trying our best to contribute to a community project here :)

File: tests/test_api.py (hunk 2 above)
Comment (samj1912) [reply tests/test_api.py]: I also updated these links to point the the right repo. :) Thanks for the PR with the instructions.

File: cruft/_commands/utils/generate.py (hunk 10 above)
Comment (gilbsgilbs) [reply cruft/_commands/utils/generate.py:83]: I appreciate your reply and I admit that I may have slightly overreacted. What annoyed me was that you seemed to imply and insisted multiple times in the same message that I changed the code for no reason while I tried to refactor it in a way makes sense overall. I had to deeply refactor the `cookiecutter_template` function anyways because updating and diffing were tightly coupled. So that's very different from the 3-4 lines I changed in tests that indeed had nothing to deal with the content of this PR. (By the way, for the latter, I know some maintainers like and even encourage single-line trivial improvements, but I couldn't know you didn't like it before making the PR :smile: ).

Now that you gave a clear and objective statement on what was the problem with this change, I'm just fine with it and I can look for solutions. IMO your first message lacked strong, verifiable, objective arguments which was later fixed by your reply. That's why I found it unfriendly and unwelcoming.

> Lastly, I think your reminder also applies to me, this is not my full time job and I'm doing my best to review this as quickly as possible in my spare time on a Sunday :)

Yup, totally applies to you (that's why I said “also”) and I am very sorry if I made this PR harder to review for you than it should have. I'll try my best to improve that in my further contributions to this project. Most importantly, I do appreciate and am very grateful for your contribution to the open-source community. Cruft is a great and well-thought project. I really hope it'll get the visibility it deserves.

File: cruft/_cli.py (hunk 11 above)
Comment (gilbsgilbs) [reply cruft/_cli.py:244]: Sorry for the sarcastic comment. Didn't know it wasn't in line with the CoC.

File: cruft/_commands/utils/generate.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+def cookiecutter_template(
+    output_dir: Path,
+    repo: Repo,
+    project_dir: Path = Path("."),
+    cookiecutter_input: bool = False,
+    checkout: Optional[str] = None,
+) -> CookiecutterContext:
+    """Generate a clean cookiecutter template in output_dir."""
+    pyproject_file = project_dir / "pyproject.toml"
+    cruft_file = get_cruft_file(project_dir)
```
Comment (samj1912) [inline cruft/_commands/utils/generate.py:31]: Would it be possible to add a cruft state argument to the function and pass it directly? Otherwise this currently leads to the cruft state being read twice - once in the relevant CLI commands and once here.

Comment (gilbsgilbs) [reply cruft/_commands/utils/generate.py]: Yup, nice catch actually :monocle_face: . Fixed it.

**Comment by gilbsgilbs** [comment]
Thanks for updating the links :+1:

File: cruft/_commands/utils/generate.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+        # Kindly ask cookiecutter to generate the template
+        template_dir = generate_files(
+            repo_dir=inner_dir, context=new_context, overwrite_if_exists=True, output_dir=tmpdir
+        )
+        template_dir = Path(template_dir)
+
+        # Move the template content to the output directory
+        for name in os.listdir(template_dir):
+            move(str(template_dir / name), str(output_dir))
```
Comment (samj1912) [inline cruft/_commands/utils/generate.py:81]: IIUC, this might fail if the generated folder has a subfolder with the same name?

For eg, what would happen if template_dir looked like -

```
test
└── test
    └── asd
--
```

Where output_dir is on the same level as the top level `test` dir.

Comment (gilbsgilbs) [reply cruft/_commands/utils/generate.py:81]: I may be missing something but I don't see why it would. In you exemple:
- the loop would make one iteration because `os.listdir` is shallow and `template_dir` contains one entry (the inner `test` directory)
- `template_dir` would equal, say, `/tmpdir/xxx/test`
- `name` would equal `test`
- `output_dir` would equal, say, `/tmpdir/yyy`

So at the end of the day, one call would be issued to the move function and this call would be equivalent to `shutil.move("/tmpdir/xxx/test/test", "/tmpdir/yyy")` which, as far as I understand, is what we want. The inner `test` directory would be moved to the `output_dir`.

Does that make sense? Am I missing something?

Comment (gilbsgilbs) [reply cruft/_commands/utils/generate.py:81]: Oh, I think I see what you meant. The `_générateur_template` function expects the `output_dir` to exist which is gross. I'll move the mkdir call in this function as soon as I can reach a computer.

Comment (gilbsgilbs) [reply cruft/_commands/utils/generate.py:81]: Does [this](https://github.com/cruft/cruft/pull/59/commits/d772a9d02f7da16e4d3b2cd8662b222ff0054c73) address your remark?

**Comment by samj1912** [comment]
So I have been thinking about this PR and I have a couple more thoughts on how this can be used to make `cruft update` seamless. Now that we have the ability to diff the local version of the project from the upstream template that created it - it would be nice if we have a way to apply this diff back to the project somehow. I have also been thinking that `cruft diff` might be a useful check to run as a part of `cruft update` where it could warn users about files that have diverged and ask users to either add the diverged files to cruft.skip or reset them to their original state before applying the `update` so that it goes smoothly. This does not necessarily have to be a part of this PR but that may change the way code has been organized in this PR. @gilbsgilbs what are your thoughts on this? Do you think we need any changes to the current PR given the above use case is now a possibility? Otherwise I will take a final look at the PR tonight and merge it in if it looks good :)

**Comment by gilbsgilbs** [comment]
Hi! Thanks for reaching out.

I'd argue to move one step at a time and I'd suggest to make extra-refactoring in a dedicated PR (as the goal is not clearly defined yet).

You made a good point though: this is exactly how I plan on using cruft and I suspect I'm not the only one. Still, I feel that the current behavior of the update command has some upsides too. A good example of this is the `pyproject.toml`: as this file may contain your project name, it may have diverged from the remote template but you neither want to skip it nor get the full diff prompted each time you update. That's even a limitation of the diff command I thought of while working on this PR: in an ideal world we'd want separate skips for updated files and diffed files (but that'd be weird to have two "skip", that's why I chose to not make it).

I'm not sure what'd be the best way to address all use-cases, but IMHO we'll need to choose the wording and expected behavior carefully if we don't want to make the CLI and cruft itself more cumbersome and complex than it should be.

**Review by samj1912 — APPROVED** [review]

**Comment by samj1912** [comment]
Thanks for the PR! Looking forward to the next set of PRs!

**Comment by gilbsgilbs** [comment]
Thanks for the thoughtful review :pray: .
